gui/src/app/tinystan/index.ts

```diff
@@ -386,6 +446,115 @@ export default class StanModel {
     });
   }
 
+  public pathfinder(p: Partial<PathfinderParams>): StanDraws {
+    const {
+      data,
+      num_paths,
+      inits,
+      seed,
+      id,
+      init_radius,
+      num_draws,
+      max_history_size,
+      init_alpha,
+      tol_obj,
+      tol_rel_obj,
+      tol_grad,
+      tol_rel_grad,
+      tol_param,
+      num_iterations,
+      num_elbo_draws,
+      num_multi_draws,
+      calculate_lp,
+      psis_resample,
+      refresh,
+      num_threads,
+    } = { ...defaultPathfinderParams, ...p };
+
+    if (num_paths < 1) {
+      throw new Error("num_paths must be at least 1");
+    }
+    if (num_draws < 1) {
+      throw new Error("num_draws must be at least 1");
+    }
+    if (num_multi_draws < 1) {
+      throw new Error("num_multi_draws must be at least 1");
+    }
+
+    const output_rows =
+      calculate_lp && psis_resample ? num_multi_draws : num_draws * num_paths;
+
+    const seed_ = seed !== null ? seed : Math.floor(Math.random() * (2 ^ 32));
+
+    return this.withModel(data, seed_, (model, deferredFree) => {
+      const rawParamNames = this.m.UTF8ToString(
+        this.m._tinystan_model_param_names(model),
+      );
+      const paramNames = PATHFINDER_VARIABLES.concat(rawParamNames.split(","));
+
+      const n_params = paramNames.length;
+
+      const free_params = this.m._tinystan_model_num_free_params(model);
+      if (free_params === 0) {
+        throw new Error("Model has no parameters.");
+      }
+
+      const inits_ptr = this.encodeInits(inits);
+      deferredFree(inits_ptr);
+
+      const n_out = output_rows * n_params;
+      const out = this.m._malloc(n_out * Float64Array.BYTES_PER_ELEMENT);
+      deferredFree(out);
+      const err_ptr = this.m._malloc(PTR_SIZE);
+      deferredFree(err_ptr);
+
+      const result = this.m._tinystan_pathfinder(
+        model,
+        num_paths,
+        inits_ptr,
+        seed_ || 0,
+        id,
+        init_radius,
+        num_draws,
+        max_history_size,
+        init_alpha,
+        tol_obj,
+        tol_rel_obj,
+        tol_grad,
+        tol_rel_grad,
+        tol_param,
+        num_iterations,
+        num_elbo_draws,
+        num_multi_draws,
+        calculate_lp ? 1 : 0,
+        psis_resample ? 1 : 0,
+        refresh,
+        num_threads,
+        out,
+        n_out,
+        err_ptr,
+      );
+
+      if (result != 0) {
+        this.handleError(err_ptr);
+      }
+
+      const out_buffer = this.m.HEAPF64.subarray(
+        out / Float64Array.BYTES_PER_ELEMENT,
+        out / Float64Array.BYTES_PER_ELEMENT + n_out,
+      );
+
+      const draws: number[][] = Array.from({ length: n_params }, (_, i) =>
+        Array.from(
+          { length: output_rows },
+          (_, j) => out_buffer[i + n_params * j],
+        ),
+      );
+
+      return { paramNames, draws };
+    });
+  }
+
```

Comment:
Might want to convert the `_malloc(4)` calls below to use `PTR_SIZE` instead, since you added it.

Reply:
the 4 in those mallocs is a different 4 than the ones I changed here (it is not necessarily true that `sizeof(int*) == sizeof(int)`, in fact on x64 systems it is false)

So I could define an INT_SIZE const variable, but those are also the only places I think the API will ever need to malloc an int (in retrospect, I could use `Int32Array.BYTES_PER_ELEMENT` as something more self-documenting, but also the fact that `Int32` is in the name does pretty strongly imply the answer will be 4...)